### ISSUE TITLE
Feature request: maxTimeout option for replication retry

### ISSUE THREAD

**Issue by colinskow** [pouchdb/pouchdb#5130]
If my internet goes down, I want PouchDB to reconnect as soon as practically possible. I don't want the backoff function to get to the point where it is waiting hours, days, or weeks.

I would like to add `options.maxTimeout` to `PouchDB.replicate`. This will ensure the retry delay never exceeds the value specified. If the delay exceeds `maxTimeout` then the default backoff function will choose a random delay between 50% of `maxTimeout` and `maxTimeout`.

From: `src/replicate/backoff.js`

```js
// Backoff function from PouchDB
  // Starts with a random number between 0 and 2 seconds and doubles it after every failed connect
  // Will not go higher than options.maxTimeout
  function randomNumber(min, max) {
    min = parseInt(min, 10) || 0;
    max = parseInt(max, 10);
    if (max !== max || max <= min) {
      max = (min || 1) << 1; //doubling
    } else {
      max = max + 1;
    }
    // In order to not exceed maxTimeout, pick a random value between 50% of maxTimeout and maxTimeout
    if(options.maxTimeout && max > options.maxTimeout) {
      min = options.maxTimeout >> 1; // divide by two
      max = options.maxTimeout;
    }
    var ratio = Math.random();
    var range = max - min;

    return ~~(range * ratio + min); // ~~ coerces to an int, but fast.
  }
```

I can do a pull request including tests and documentation, but wanted to discuss the idea here first.

**Comment by nolanlawson** [comment]
PouchDB actually allows the user to specify custom backoff functions. The option is described [here](https://pouchdb.com/api.html#replication) under `back_off_function`. Using this method, it seems like you should be able to describe the timeout you describe here. Does that work for you?

**Comment by colinskow** [comment]
The custom back_off_function works, but...
1) It is a lot of boilerplate to duplicate and modify your default function
2) A maximum timeout is a practical feature that will benefit a lot of users. It doesn't make much sense to have the timeout expand infinitely.
3) Since `maxTimeout` is an optional option it won't affect existing applications unless they choose to use it

**Comment by nolanlawson** [comment]
The point of the `back_off_function` option was avoid having to support every possible retry strategy, and just allow users to roll their own if they needed something more specialized than our implementation. The algorithm you describe could definitely be implemented as a separate plugin/module.

OTOH I agree that a `maxTimeout` makes sense. Having PouchDB wait hours or days to make a new request seems silly. I'd be in favor of hard-coding a `maxTimeout` into the existing algorithm, but I don't think it should be configurable – rather, users should just override `back_off_function` entirely if they want to change the `maxTimeout`. This gives us one less thing to have to support at the API level.

**Comment by colinskow** [comment]
I understand not wanting to add more complexity to the API. In that case we need to decide on a practical value to hard code.

A connection attempt really doesn't cost anything, but sending one every few seconds is probably unacceptable. To me 5-10 minutes seems practical in the real world. So once the delay goes over 5 minutes it will randomly pick a value between 5 and 10 minutes.

**Comment by daleharvey** [comment]
So I thought we already implemented an upper limit on how long our retries would run, but looking at the code it seems I was wrong. So yeh I would be +1 on having a max timeout

**Comment by colinskow** [comment]
@daleharvey should we hard-code it as @nolanlawson suggests, or leave it modifiable by the user?

Should we keep the default behavior as it is, or set it to a sensible 10 minute maxTimeout? If the internet is out for 1 hour, I don't think most users will want to wait 2 hours to reconnect.

**Comment by daleharvey** [comment]
So I agree with Nolan that any where we can reduce / avoid adding to the API is a good thing, but at the same time this seems like a single addition that should be fairly simple, avoids potential future changes and will also make things slightly easier to test, @nolanlawson how much are you in against adding? I would say I am 60/40 for.

**Comment by colinskow** [comment]
For me the current setup of 2x exponential backoff with randomness is perfect... IF I can control the maximum delay. Ten minutes seems like a sane value for the majority of use cases. But being able to tweak this number I can't foresee any reason to override the default function, which is a chunk of code and took me a while to decipher.

So my proposal is:
`options.maxTimeout`
Default value 600000 (10 minutes).
0 will enable infinite backoff.

Anyone opposed speak up now.

**Comment by colinskow** [comment]
Also @nolanlawson is it safe to PR now or should I wait for monorepo?

**Comment by nolanlawson** [comment]
@daleharvey I would much rather just set a default `maxTimeout` of 10 minutes as suggested by @colinskow and then not expose it as a new API. There are many hard-coded values in the built-in retry replication algorithm; I think exposing a single one of them as a new API introduces unnecessary complexity (how would `back_off_function` and `maxTimeout` interact?), especially given that users can just override the whole thing with `back_off_function`, and we can still do that in tests to make it more testable.

So in summary I'm for adding a hardcoded timeout of 10 minutes as suggested by @colinskow , but not for adding a user-facing `maxTimeout` option.

@colinskow Feel free to open a PR right now; I think this change would be pretty easy to merge with my monorepo work.

**Comment by daleharvey** [comment]
And fixed in https://github.com/pouchdb/pouchdb/commit/18abf466e7171e18cc6a90fed0cc883026a826d7
